docs/spec/distributing.rst

```diff
@@ -10,72 +10,156 @@ Stub files
 
 (Originally specified in :pep:`484`.)
 
-Stub files are files containing type hints that are only for use by
-the type checker, not at runtime.  There are several use cases for
-stub files:
+*Stub files*, also called *type stubs*, provide type information for untyped
+Python packages and modules. Stub files serve multiple purposes:
+
+* They are the only way to add type information to extension modules.
+* They can provide type information for packages that do not wish to
+  add them inline.
+* They can be distributed separately from the package or module that they
+  provide types for. The latter is referred to as the *implementation*.
+  This allows stubs to be developed at a different pace or by different
+  authors, which is especially useful when adding type annotations to
+  existing packages.
+* They can act as documentation, succinctly explaining the external
+  API of a package, without including implementation details or private
+  members.
+
+Stub files use a subset of the constructs used in Python source files, as
+described in :ref:`stub-file-supported-constructs` below. Type checkers should
+parse a stub that uses only such constructs without error and not interpret any
+construct in a manner contradictory to this specification. However, type
+checkers are not required to implement checks for all of these constructs and
+can elect to ignore unsupported ones. Additionally, type checkers can support
+constructs not described here.
+
+If a stub file is found for a module, the type checker should not read the
+corresponding "real" module. See :ref:`mro` for more information.
+
+.. _stub-file-syntax:
+
+Syntax
+^^^^^^
+
+Stub files are syntactically valid Python files in the earliest Python version
+that is not yet end-of-life. They use a ``.pyi`` suffix. The Python syntax used
+by stub files is independent from the Python versions supported by the
+implementation, and from the Python version the type checker runs under (if
+any).
+
+Type checkers should evaluate all annotation expressions as if they are quoted.
+Consequently, forward references do not need to be quoted, and syntax from newer
+versions than otherwise supported may be used in annotations. For example, the
+pipe union syntax (``X | Y``) introduced in Python 3.10 may be used even before
+Python 3.9 reaches end-of-life. Outside of annotations, type checkers may but
+are not required to support type system features that use newer syntax.
```

Comment:
I think this text is still somewhat confusing, in particular because it uses "syntax" twice in different meanings. I would restrict "syntax" to what the parser looks at, so the ``|`` union operator is not "syntax". I tried to make the text more explicit about the requirement: stubs should be *parsable* in all Python versions, but once the code is parsed, it may use features that do not exist in the earliest supported Python version.

```suggestion
Stub files are syntactically valid Python files with a ``.pyi`` suffix. They
should be parseable (e.g., with :py:func:`ast.parse`) in all Python versions
supported by the implementation. For example, defining a type alias using the
``type`` keyword is only accepted by the Python parser in Python 3.12 and later,
so stubs supporting Python 3.11 or earlier versions should not use this syntax.
Type checkers may choose to support syntactic features from newer Python versions
in stub files, but stubs that rely on such features may not be portable to all
type checkers.

Type checkers should evaluate all :term:`annotation expressions <annotation expression>` as if they are quoted.
Consequently, forward references do not need to be quoted, and type system
features that do not depend on Python syntax changes are supported in stubs regardless
of the Python version supported. For example, the use of the ``|`` operator
to create unions (``X | Y``) was introduced in Python 3.10, but may be used
even in stubs that support Python 3.9 and older versions. 
```

This I believe accurately describes the current state of stubs in e.g. typeshed. Restricting stubs to what can be parsed by `ast.parse` allows type checkers that rely on `ast.parse` (mypy and pyanalyze) to work and more generally makes it possible for tools that are written in Python to analyze stub files without having to build their own parser. I think that's useful and desirable.

Reply:
I think this phrasing is better than previous drafts, but I'm still uneasy with it. I again question whether guidance to stub authors (what they should or should not put in the stubs) belongs in the typing spec. The typing spec is meant to specify the required and optional behavior for type checkers. From that perspective, the new sentence that begins "Type checkers may choose..." seems appropriate for the typing spec, but the sentence preceding it seems out of place. Is there a way it can be rephrased such that it indicates what is expected of type checkers?

In the phrase "in all Python versions supported by the implementation", what does "implementation" mean? Is this referring to the type checker implementation or the implementation of the library associated with the stub? If it's the latter, that would seem to imply that a library that supports Python 2.x wouldn't be able to use any Python 3.x syntax in its corresponding stub.

How about something like "A type checker implemented in Python may depend upon the Python parser (e.g., with :py:func:`ast.parse`) to parse source and stub files. If this type checker runs on an older version of Python, it will not be able to parse newer syntactic constructs. For example, a type alias defined with the `type`` keyword is accepted by the Python parser starting in Python 3.12, so a stub that uses this syntax may not be supported by a type checker running on Python 3.11 or older. For further guidance about authoring stubs, refer to <link to guide>".

Reply:
> The typing spec is meant to specify the required and optional behavior for type checkers. 

In PEP 729, we wrote that type checkers are one audience for the spec, but not the only one: https://peps.python.org/pep-0729/#specification-for-the-type-system. I see this section as specifying what is required from both type checkers and stub authors. On the one hand, we have to specify what type checkers need to support in order to understand a conformant stub; on the other hand, we have to specify what features stub authors can use when they write stubs in order to be conformant.

From that perspective, I actually feel that the "Type checkers may choose..." sentence is redundant. Type checkers may always choose to support more than what the spec requires, but the spec should be precise about the minimum required, so that stub authors can know what to rely on.

> what does "implementation" mean?

An earlier paragraph in this document defines "implementation" (and it means what you guessed). But you make a good point about Python 2; we shouldn't require Python 2 syntax. A good general solution would be to require that stubs are parseable only in Python version that are still supported by CPython (i.e., currently 3.8 through 3.13).

> A type checker implemented in Python may depend upon the Python parser (e.g., with :py:func:ast.parse) to parse source and stub files

This feels more like a motivation than a specification. The spec should primarily say what behavior is required, not why it is that way, though we can add explanatory text where it is helpful.

With that in mind here's another suggested text:

```suggestion
Stub files are syntactically valid Python files with a ``.pyi`` suffix. They
should be parseable (e.g., with :py:func:`ast.parse`) in all Python versions
that are supported by the implementation and that are still supported
by the CPython project. For example, defining a type alias using the
``type`` keyword is only accepted by the Python parser in Python 3.12 and later,
so stubs supporting Python 3.11 or earlier versions should not use this syntax.
This allows type checkers implemented in Python to parse stub files using
functionality from the standard library.
Type checkers may choose to support syntactic features from newer Python versions
in stub files, but stubs that rely on such features may not be portable to all
type checkers. Type checkers may also choose to support Python versions that
are no longer supported by CPython; if so, they cannot rely on standard library
functionality to parse stub files.

Type checkers should evaluate all :term:`annotation expressions <annotation expression>` as if they are quoted.
Consequently, forward references do not need to be quoted, and type system
features that do not depend on Python syntax changes are supported in stubs regardless
of the Python version supported. For example, the use of the ``|`` operator
to create unions (``X | Y``) was introduced in Python 3.10, but may be used
even in stubs that support Python 3.9 and older versions. 
```

Reply:
This phrasing still makes me uneasy, but I think its meaning is now clear and it represents a reasonable compromise from where we started. Thanks for listening to my concerns.

Reply:
@JelleZijlstra Thanks, I think this is a good compromise, and the wording is clearer than before.